File: tests/fixtures/quality/python/service.py
"""Coordinates background jobs and their retry policy."""

from dataclasses import dataclass
from typing import Optional

MAX_ATTEMPTS = 5
_DEFAULT_QUEUE = "jobs"


class Runnable:
    """Anything the scheduler can drive."""

    def start(self) -> None:
        """Begin processing."""
        raise NotImplementedError


@dataclass
class JobSpec:
    """A unit of deferred work."""

    name: str
    attempts: int = 0


class Scheduler(Runnable):
    """Runs jobs with exponential backoff between retries."""

    def __init__(self, queue: str = _DEFAULT_QUEUE) -> None:
        self.queue = queue
        self._pending: list[JobSpec] = []

    def start(self) -> None:
        """Drain the pending queue."""
        for spec in self._pending:
            self.dispatch(spec)

    def dispatch(
        self,
        spec: JobSpec,
        timeout: Optional[float] = None,
    ) -> bool:
        """Run one job, retrying with exponential backoff."""

        def next_delay(attempt: int) -> float:
            return 2.0**attempt

        while spec.attempts < MAX_ATTEMPTS:
            next_delay(spec.attempts)
            spec.attempts += 1
        return True

    def _reset(self) -> None:
        self._pending.clear()


async def drain(scheduler: Scheduler) -> None:
    """Await every in-flight job."""
    scheduler.start()

File: tests/fixtures/quality/go/service.go
// Package worker runs queued jobs with a retry budget.
package worker

import "fmt"

// MaxAttempts bounds how often a job is retried.
const MaxAttempts = 5

var defaultQueue = "jobs"

// Runnable is anything the scheduler can drive.
type Runnable interface {
	// Start begins processing.
	Start() error
	Depth() int
}

// JobSpec is a unit of deferred work.
type JobSpec struct {
	// Name identifies the job.
	Name     string
	Attempts int
}

// Scheduler runs jobs with exponential backoff between retries.
type Scheduler struct {
	Queue   string
	pending []JobSpec
}

// Audited adds an audit trail to any scheduler.
type Audited struct {
	Scheduler
	Log []string
}

// Start drains the pending queue.
func (s *Scheduler) Start() error {
	for _, spec := range s.pending {
		if err := s.Dispatch(spec); err != nil {
			return err
		}
	}
	return nil
}

// Dispatch runs one job, retrying with exponential backoff.
func (s *Scheduler) Dispatch(
	spec JobSpec,
) error {
	nextDelay := func(attempt int) int { return 1 << attempt }
	for spec.Attempts < MaxAttempts {
		nextDelay(spec.Attempts)
		spec.Attempts++
	}
	return fmt.Errorf("done: %s", spec.Name)
}

func (s *Scheduler) reset() {
	s.pending = nil
}

// Depth reports the queue length.
func (s *Scheduler) Depth() int { return len(s.pending) }
